scipy/optimize/_optimize.py

```diff
@@ -358,9 +360,10 @@ def rosen(x):
     >>> ax.plot_surface(X, Y, rosen([X, Y]))
     >>> plt.show()
     """
-    x = asarray(x)
-    r = np.sum(100.0 * (x[1:] - x[:-1]**2.0)**2.0 + (1 - x[:-1])**2.0,
-                  axis=0)
+    xp = array_namespace(x)
+    x = xp.asarray(x)
+    r = xp.sum(100.0 * (x[1:] - x[:-1]**2.0)**2.0 + (1 - x[:-1])**2.0,
+               axis=0, dtype=x.dtype)
```

Comment:
I suppose it's fine if `x` is of integer dtype, since this will produce integers. Perhaps `100.0` should be `100`, then, if we don't want to cast back and forth.

Reply:
Hmm actually I think I should change the dtype here as the sum previously would always be an inexact dtype due to the `100.0`

Reply:
Also integers don't work anyway under array API strict

Reply:
Then I'd take the liberty of changing the derivatives, too.

Reply:
Done